File: 프로그래머스/lv2/42584. 주식가격/주식가격.py
def solution(prices):
    # 현재 보는 가격보다 이전가격이 큰 경우 그때 초기화됨
    # stack에 저장하면서 위치 정보와 같이 저장함
    
    s = []
    n = len(prices)
    res = [0]*n
    
    for i, p in enumerate(prices):
        idx = (n-i)-1
        while s and s[-1][0] > p:
            before_price, before_idx = s.pop()
            res[n-before_idx-1] = before_idx-idx
        s.append((p, idx))
    
    while s:
        price, idx = s.pop()
        res[n-idx-1] = idx
        
    return res

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;
import java.io.*;

class Solution {
    public int[] solution(int[] prices) {
        Deque<int[]> dq = new LinkedList<>();
        ArrayList<int[]> ps = new ArrayList<>();
        
        for(int i=0; i<prices.length; i++){
            ps.add(new int[] {i, prices[i]}); // idx, value
        }
        
        
        int res[] = new int[ps.size()];
        
        for(int i=0;i<ps.size(); i++){
            while(!dq.isEmpty() && dq.peekLast()[1] > ps.get(i)[1]){
                int idx = dq.pollLast()[0];
                res[idx] = i-idx;
            }
            dq.addLast(ps.get(i));
        }
        
        while(!dq.isEmpty()){
            int idx = dq.pollLast()[0];
            res[idx] = ps.size() - 1 - idx;
        }
        
        return res;
    }
}
